src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -274,12 +339,167 @@ public void WriteRawValue(string rawValue)
                 return;
             }
 
+            // ensure we don't write to the stream directly while there are still pending data in the Utf8JsonWriter buffer
+            this.Flush(); 
+            if (IsInArray() && !isWritingFirstElementInArray)
+            {
+                // Place a separator before the raw value if
+                // this is an array, unless this is the first item in the array.
+                this.writeStream.WriteByte(itemSeparator.Span[0]);
+            }
+
             // Consider using Utf8JsonWriter.WriteRawValue() in .NET 6+
             // see: https://github.com/OData/odata.net/issues/2420
-
-            this.Flush(); // ensure we don't write to the stream while there are still pending data in the buffer
             this.writeStream.Write(Encoding.UTF8.GetBytes(rawValue));
-            this.FlushIfBufferThresholdReached();
+
+            // since we bypass the Utf8JsonWriter, we need to signal to other
+            // Write methods that a separator should be written first
+            if ((this.isWritingFirstElementInArray || this.isWritingConsecutiveRawValuesAtStartOfArray) || !this.IsInArray())

```

Comment:
what if we writing the RawValue at the end of array or at the middle of array?

Reply:
can we have a case like this? [`NonRawValue`, `RawValue`, `NonRawValue`, `RawValue`]

Reply:
I also haven't seen the case of an object scenario. Like where is `this.shouldWriteSeparator = true;` set for object scenario?

Reply:
Good questions. All those scenarios are supported, and the payload in the test cases include such examples.

In the middle of the array, raw values are not such a big deal. I just need to make sure I place a comma before the raw value, but I don't need to worry about what comes next. This is because inside array the `Utf8JsonWriter` will automatically add a comma before each value, so we don't need to add one manually.

At the end of the array, we don't need to add separators either. So what I'm doing at the end of arrays in the `ExitScope` method, I just reset the state of the tracking variables.

I handle objects differently from arrays because in objects I just need to make sure I place the comma before property names that a follow a raw value. So you'll notice inside `WriterName` I check whether we should add a separator by checking the `this.shouldWriterSeparator` variable. And I set that variable to true inside the `WriteRawValue` method if we're *not* inside an array (i.e. if we're inside an object).

The reason I have special treatment for raw values at the start of an array is because it was the tricky case: I have to ensure that I don't add a separator before the raw value (which is a different treatment from raw values in the middle of an array), I also have to make sure that I have a separator after the raw values of the start of the array before the first non-raw value because `Utf8JsonWriter` will not add it automatically since it mistakenly thinks it's at the beginning of the array.